src/replication.c

```diff
@@ -2764,6 +2715,233 @@ cleanup_deadlist(void *arg)
+{
+	replica_t *replica;
+	struct timespec last, now, diff;
+	int ret = -1, rc = -1;
+	uint64_t io_seq = 0;
+
+	ASSERT(spec);
+	ASSERT(rcomm_mgmt_cmd);
+
+	clock_gettime(CLOCK_MONOTONIC, &last);
+	MTX_LOCK(&spec->rq_mtx);
+
+	/* Wait for any ongoing mgmt commands which could block IOs. */
+	while (spec->quiesce == 1) {
+		MTX_UNLOCK(&spec->rq_mtx);
+		sleep(1);
+		MTX_LOCK(&spec->rq_mtx);
+	}
+
+	if (is_volume_healthy(spec) == false) {
+		REPLICA_ERRLOG("volume is not healthy..\n");
+		ret = 1;
+		goto out;
+	}
+
+	if (!pause_and_timed_wait_for_ongoing_ios(spec, io_wait_time)) {
+		REPLICA_ERRLOG("pausing failed..\n");
+		ret = 1;
+		goto out;
+	}
+
+	/*
+	 * If snap destroy is called due to error occurred in snap create
+	 * then io_seq number should be same as in snap create mgmt command
+	 * otherwise it should be 0.
+	 */
+	if (opcode != ZVOL_OPCODE_SNAP_DESTROY)
+		io_seq = ++spec->io_seq;
+	else if (mgmt_io_seq && *mgmt_io_seq != 0 &&
+	    opcode == ZVOL_OPCODE_SNAP_DESTROY)
+		io_seq = *mgmt_io_seq;
+
+	if (mgmt_io_seq)
+		*mgmt_io_seq = io_seq;
+
+	MTX_LOCK(&rcomm_mgmt_cmd->mtx);
+
+	TAILQ_FOREACH(replica, &spec->rq, r_next) {
+		rc = enqueue_mgmt_command_to_replica(replica, io_seq,
+		    rcomm_mgmt_cmd->buf, rcomm_mgmt_cmd->buf_size, opcode,
+		    rcomm_mgmt_cmd);
+		if (rc < 0) {
+			rcomm_mgmt_cmd->caller_gone = 1;
+			MTX_UNLOCK(&rcomm_mgmt_cmd->mtx);
+			REPLICA_ERRLOG("caller gone..\n");
+			ret = -1;
+			goto out;
+		}
+	}
+
+	timesdiff(CLOCK_MONOTONIC, last, now, diff);
+
+	if (rcomm_mgmt_cmd->cmds_sent != spec->replication_factor) {
+		rcomm_mgmt_cmd->caller_gone = 1;
+		MTX_UNLOCK(&rcomm_mgmt_cmd->mtx);
+		REPLICA_ERRLOG("cmds sent %d not eq repl factor %d..\n",
+		    rcomm_mgmt_cmd->cmds_sent, spec->replication_factor);
+		ret = -1;
+		goto out;
+	}
+
+	while (diff.tv_sec < wait_time) {
+		if (rcomm_mgmt_cmd->cmds_sent ==
+		    (rcomm_mgmt_cmd->cmds_succeeded +
+		    rcomm_mgmt_cmd->cmds_failed))
+			break;
+		MTX_UNLOCK(&rcomm_mgmt_cmd->mtx);
+		MTX_UNLOCK(&spec->rq_mtx);
+		sleep(1);
+		MTX_LOCK(&spec->rq_mtx);
+		MTX_LOCK(&rcomm_mgmt_cmd->mtx);
+		timesdiff(CLOCK_MONOTONIC, last, now, diff);
+	}
+
+	rcomm_mgmt_cmd->caller_gone = 1;
+	if (rcomm_mgmt_cmd->cmds_sent ==
+	    (rcomm_mgmt_cmd->cmds_succeeded + rcomm_mgmt_cmd->cmds_failed)) {
+		if ((rcomm_mgmt_cmd->cmds_succeeded ==
+		    spec->replication_factor) &&
+		    (rcomm_mgmt_cmd->cmds_failed == 0)) {
+			ret = 0;
+		} else {
+			ret = -1;
+		}
+
+		MTX_UNLOCK(&rcomm_mgmt_cmd->mtx);
+		free_rcommon_mgmt_cmd(rcomm_mgmt_cmd);
+		goto out;
+	}
+
+	MTX_UNLOCK(&rcomm_mgmt_cmd->mtx);
+
+out:
+	spec->quiesce = 0;
+	MTX_UNLOCK(&spec->rq_mtx);
+	return ret;
+}
+
+int
+istgt_lu_resize(spec_t *spec, uint64_t new_size, uint64_t old_size)
+{
+	rcommon_mgmt_cmd_t *rcomm_mgmt;
+	int rc = 0, rc1 = 0;
+
+	rcomm_mgmt = allocate_rcommon_mgmt_cmd(sizeof (new_size));
+	*(uint64_t *)rcomm_mgmt->buf = new_size;
+	rcomm_mgmt->buf_size = sizeof (new_size);
```

Comment:
We have already set rcomm_mgmt->buf_size inside allocate_rcommon_mgmt_cmd so we can remove this.